File: src/0102.binary-tree-level-order-traversal.102/solution.go
package binarytreelevelordertraversal102

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func levelOrder(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	result := [][]int{}
	var bfs func([]*TreeNode)
	bfs = func(levelNodes []*TreeNode) {
		vals := []int{}
		next := []*TreeNode{}
		for _, node := range levelNodes {
			vals = append(vals, node.Val)
			if node.Left != nil {
				next = append(next, node.Left)
			}
			if node.Right != nil {
				next = append(next, node.Right)
			}
		}
		result = append(result, vals)
		if len(next) > 0 {
			bfs(next)
		}
	}
	bfs([]*TreeNode{root})
	return result
}

File: src/0102.binary-tree-level-order-traversal.102/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        List<List<Integer>> result = new ArrayList<>();

        travel(List.of(root), result);
        return result;
    }

    private void travel(List<TreeNode> level, List<List<Integer>> result) {
        List<Integer> row = new ArrayList<>();
        List<TreeNode> nextLevel = new ArrayList<>();
        for (TreeNode node : level) {
            row.add(node.val);
            if (node.left != null) {
                nextLevel.add(node.left);
            }
            if (node.right != null) {
                nextLevel.add(node.right);
            }
        }
        result.add(row);
        if (!nextLevel.isEmpty()) {
            travel(nextLevel, result);
        }

    }

}